segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java

```diff
@@ -815,6 +821,62 @@ public void readTableEntries(WireCommands.ReadTableEntries readTableEntries) {
                 }).exceptionally(e -> handleException(readTableEntries.getRequestId(), segment, operation, e));
     }
 
+    @Override
+    public void readTableEntriesDelta(WireCommands.ReadTableEntriesDelta readTableEntriesDelta) {
+        final String segment = readTableEntriesDelta.getSegment();
+        final String operation = "readTableEntriesDelta";
+
+        if (!verifyToken(segment, readTableEntriesDelta.getRequestId(), readTableEntriesDelta.getDelegationToken(), operation)) {
+            return;
+        }
+
+        final int suggestedEntryCount = readTableEntriesDelta.getSuggestedEntryCount();
+        final long fromPosition = readTableEntriesDelta.getFromPosition();
+
+        log.info(readTableEntriesDelta.getRequestId(), "Fetching keys from {}.", readTableEntriesDelta);
+
+        val timer = new Timer();
+        val result = new DeltaIteratorResult<BufferView, Map.Entry<WireCommands.TableKey, WireCommands.TableValue>, DeltaIteratorState>(
+                segment.getBytes().length + WireCommands.TableEntriesRead.HEADER_BYTES);
+        tableStore.entryDeltaIterator(segment, fromPosition, TIMEOUT)
+                .thenCompose(itr -> itr.collectRemaining(
+                        e -> {
+                            if (result.getItemCount() >= suggestedEntryCount || result.getSizeBytes() >= MAX_READ_SIZE) {
+                                return  false;
+                            }
+                            TableEntry entry = e.getEntries().iterator().next();
+                            DeltaIteratorState state = DeltaIteratorState.deserialize(e.getState().array());
+                            // Store all TableEntries.
+                            val k = new WireCommands.TableKey(toByteBuf(entry.getKey().getKey()), entry.getKey().getVersion());
+                            val v = new WireCommands.TableValue(toByteBuf(entry.getValue()));
+                            if (state.isDeletionRecord()) {
+                                result.remove(entry.getKey().getKey(), k.size() + v.size());
+                            } else {
+                                Map.Entry<WireCommands.TableKey, WireCommands.TableValue> old = result.getItem(entry.getKey().getKey());
```

Comment:
This complex calculation could be simplified in the following way:
1. Just invoke `.add` on the `DeltaIteratorState`.
2. Inside `DeltaIteratorState.add`, when you invoke `put` on the map, that method will actually return the existing (replaced) value, or null if no such value existed. If you get such a value, then subtract its size from the size tracker.

Reply:
Do you mean `DeltaIteratorResult`? From my understanding it is not guaranteed that entries are read monotonically, i.e we could receive some value later on in the iteration, with an earlier version. Assuming that is true, we'd have to check these conditions prior to the `put`.  Certainly I can place that logic within the `add` call, if losing the generic functionality is acceptable.

Reply:
I am talking about the HashMap internal to `DeltaIteratorResult`. That method is synchronized so you don't have to worry about concurrency. Actually you don't need to worry about concurrency at all in this case; we only need synchronized because we pass this between threads (in sequence).

If you use the HashMap.put while synchronized you are guaranteed to get the previous value assigned to that key. It will make your code simpler and (slightly) faster.

Reply:
Ok, nevermind. I completely missed the second part of the `if` condition, where you are comparing the versions. Then what you are doing is good.